internals-js/src/utils.ts

```diff
@@ -29,6 +29,88 @@ export class MultiMap<K, V> extends Map<K, V[]> {
   }
 }
 
+/**
+ * Generic OrderedMap class that can sort keys based on an arbitrary sorting function
+ * Insert time is O(log(N))
+ * Remove is not implemented, but the trivial implementation would be O(N)
+ * Uses '<' '>' sorting by default
+ * Collisions are fine, it will just overwrite the old value
+ */
+export class OrderedMap<K,V> {
+  private _keys: K[] = [];
+  private _values: Map<K,V> = new Map<K,V>();
+  private _compareFn: (a: K, b: K) => number;
+
+  private static defaultCompareFn<K>(a: K, b: K) {
+    if (a < b) {
+      return -1;
+    } else if (b < a) {
+      return 1;
+    }
+    return 0;
+  }
+
+  constructor(compareFn: (a: K, b: K) => number = OrderedMap.defaultCompareFn) {
+    this._compareFn = compareFn;
+  }
+
+  add(key: K, value: V) {
+    if (!this._values.has(key)) {
+      this.insertKeyInOrder(key);
+    }
+    this._values.set(key, value);
+  }
+
+  get(key: K): V | undefined {
+    return this._values.get(key);
+  }
+
+  has(key: K): boolean {
+    return this._values.has(key);
+  }
+
+  size(): number {
```

Comment:
Nitpick: we may want to make this a getter to mimick the `Map` api.

Reply:
Good call, I'll change it.